automation/src/test/scala/org/broadinstitute/dsde/firecloud/fixture/DownloadFixtures.scala

```diff
@@ -0,0 +1,96 @@
+    val permissions = Set(PosixFilePermission.OWNER_WRITE, PosixFilePermission.GROUP_WRITE, PosixFilePermission.OTHERS_WRITE)
+    Files.setPosixFilePermissions(downloadPath, permissions.asJava)
+    downloadPath.toString
+     */
+
+    val downloadPath = s"chrome/downloads/${UUID.randomUUID()}"
+    val dir = new File(downloadPath)
+    dir.deleteOnExit()
+    dir.mkdirs()
+    val path = dir.toPath
+    logger.info(s"mkdir: $path")
+    val permissions = Set(
+      PosixFilePermission.OWNER_WRITE, PosixFilePermission.OWNER_READ, PosixFilePermission.OWNER_EXECUTE,
+      PosixFilePermission.GROUP_WRITE, PosixFilePermission.GROUP_READ, PosixFilePermission.GROUP_EXECUTE,
+      PosixFilePermission.OTHERS_WRITE, PosixFilePermission.OTHERS_READ, PosixFilePermission.OTHERS_EXECUTE)
+    import scala.collection.JavaConverters._
+    Files.setPosixFilePermissions(path, permissions.asJava)
+    path.toString
+  }
+
+  /**
+    * Downloads the metadata currently being viewed.
+    *
+    * If downloadPath is given, the file is given a timestamped name and moved from that location
+    * into the "downloads" directory off the current working directory. This serves two purposes:
+    *
+    * 1. Archiving the file for later inspection when tests fail
+    * 2. Keeping the browser download directory clean so that it doesn't auto-rename subsequent
+    * downloads with the same filename
+    *
+    * @param downloadPath the directory where the browser saves downloaded files
+    * @return the relative path to the moved download file, or None if downloadPath was not given
+    */
+  def downloadFile(downloadPath: String, fileName: String, downloadThing: Either[Clickable, CssSelectorQuery])(implicit webDriver: WebDriver): Option[String] = synchronized {
+
+    def archiveDownloadedFile(sourcePath: String): String = {
+      // wait up to 10 seconds for file exist
+      val f = new File(sourcePath)
+      eventually {
+        assert(f.exists(), s"Timed out (10 seconds) waiting for file $f")
+      }
+
+      val date = DateTimeFormatter.ofPattern("HH:mm:ss:N").format(LocalDateTime.now())
+      val destFile = new File(sourcePath).getName + s".$date"
+      val destPath = s"downloads/$destFile"
+      Util.moveFile(sourcePath, destPath)
+      logger.info(s"Moved file. sourcePath: $sourcePath, destPath: $destPath")
+      destPath
+    }
+
+    /*
+     * Downloading a file will open another window while the download is in progress and
+     * automatically close it when the download is complete.
+     */
+    // await condition (windowHandles.size == 1, 30)
+    // .submit call takess care waiting for a new window
+
+    downloadThing match {
+      case Left(clickable) => clickable.doClick()
+      case Right(query) => {
+        logger.info(s"form: ${query.queryString}")
+        find(query).get.underlying.submit()
+      }
+    }
+
+    for {
+      hello <- Option("why does this work???")
+    } yield archiveDownloadedFile(s"$downloadPath/$fileName")
```

Comment:
So `archiveDownloadedFile(s"$downloadPath/$fileName")` without a for/yield, by itself causes the following error: `Expression of type String doesn't conform to expected type TypeParameterForSyntheticFunction_`
However, when its returned as part of a for comprehension, like it is now, it works fine.

I don't understand why this is!

Reply:
The signature for `def downloadFile` says it should return an `Option[String]`. `archiveDownloadedFile` returns a `String`, so that doesn't match.

For-comps are syntactic sugar for a series of map and flatMap, so what you're doing here is equivalent to this:

```
Option("why does this work???") map { _ =>
    archiveDownloadedFile(s"$downloadPath/$fileName")
}
```

i.e. the thing in the `yield` ends up getting "boxed" into a monad of the first type the for-comp unpacks.

In any case, if you return `Option(archiveDownloadedFile(s"$downloadPath/$fileName"))` it should work fine.

I don't know why it says things about `TypeParameterForSyntheticFunction_`; that bit looks like [some IntelliJ specific thing](https://github.com/scalatest/intellij-scala-old/blob/master/src/org/jetbrains/plugins/scala/util/ScalaUtils.java#L53). It's _possible_ you might get a different (and more comprehensible) compile error if you run scalac directly here.

Reply:
Thanks! I'm removing Option altogether and just returning String